### ISSUE TITLE
consider some way to verify target OS compatibility with `install-to-filesystem`

### ISSUE THREAD

**Issue by cgwalters** [containers/bootc#192]
I got caught out with using `install-to-filesystem --replace=alongside` against a Fedora Cloud instance (using btrfs) when targeting Fedora ELN (doesn't enable btrfs).

This problem generalizes to things like installing to LVM but not having LVM in the container, etc.

This is quite tricky to test, but I think we could at least do a dirty sanity check for having the kmod in the target kernel or so?

**Comment by cgwalters** [comment]
I briefly looked at this recently, and what I *think* we could do that should work is to use `modprobe --dry-run -S $kver_in_container` with the target rootfs type.

**Comment by cgwalters** [comment]
What I just hit that's related to this is I was trying to do a `to-existing-root` install of RHEL9 on a RHEL10 target host, and that actually failed in the grub probing because the rhel9 grub failed to parse the rhel10 host filesystem layout. But of course, I think the *kernel* would also fail to parse (probably?)

**Comment by cgwalters** [comment]
Thinking about this more...I think we should probably add a stern warning if we detect major OS version skew between the host and the target. Though doing this kind of thing will clash a bit with people wanting to change `ID` and `VERSION_ID` for their custom OS builds.

**Comment by andrewdunndev** [comment]
We hit this class of problem building custom CentOS Stream 10 bootc images with additional kernel modules compiled from source. When the target disk's filesystem requires a module that isn't present in the container's kernel, `bootc install to-disk` fails deep in the process with an error that doesn't point to the root cause.

The `modprobe --dry-run -S $kver $rootfs_type` approach from your March comment seems like the right pre-flight check. It would catch the mismatch early with a clear error message before any partitioning happens.

I'd be happy to take a crack at implementing this if the approach still looks right. I'm thinking it would go in the install path after filesystem detection but before partitioning, something like:

```
modprobe --dry-run -S <container_kver> <detected_rootfs_type>
```

With a clear error on failure: "Target filesystem requires kernel module `{module}` which is not available in the container's kernel `{kver}`."

Would that be welcome?

**Comment by cgwalters** [comment]
Yeah, but I think doing this properly probably wants https://github.com/bootc-dev/bootc/issues/640 to make it fully generic
